### DISCUSSION TITLE
OB分布式事务的提交时间戳

### DISCUSSION THREAD

**Discussion by luooofan** [oceanbase/oceanbase#844] [Classroom] [answered]
看源码了解到OB分布式事务的提交时间戳获取逻辑如下：
各参与者处理 preapre 请求时：
- 第一个参与者取 gts 作为 prepare version
- 其他参与者取 gts 本地缓存作为 prepare version

各参与者将各自的 prepare version 发送给协调者

协调者取最大的 prepare version 作为事务的 commit version

有一个问题是：第一个参与者取 gts 作为 prepare version 这个操作有必要性吗，可以只从本地缓存中取吗？

我个人的理解是：从本地缓存中取相当于是给一个当前要提交的事务一个较旧的提交时间戳，虽然这个时间戳较旧，但是依然保证了读写依赖、写读依赖、写写依赖，满足并发控制要求，而且由于事务开始/语句开始时取快照版本仍要取gts（当前读），也仍能保持外部一致性

**Comment by ScienJus** [comment, marked as answer]
不是 OB 的官方开发人员，个人观点：

必须在 prepare 时现取 gts 的原因，可能和 tidb [这篇文章](https://pingcap.com/zh/blog/async-commit-principle)中「因果一致性」的例子比较像。这里举两个时序作为例子。

第一个时序：

1. A 开启事务并修改数据
2. A 提交事务
3. A 通知 B 启动
4. B 开始事务并修改数据
5. B 提交事务

因为存在 A 通知 B 这个因果关系，所以要保证 A 的 commit ts 一定要小于 B 的 commit ts。而在 A 和 B 涉及的参与者完全不同的情况下，需要额外取一个 gts 为这两个操作定序。理论上在这个时序下，获取的 gts 只要在开启事务到提交事务之间就可以，也就可以复用 start ts（aka read ts/snapshot ts），或者是本地缓存的 max gts（因为它一定大于等于 start ts）。

上面这个时序应该对应您说的「而且由于事务开始/语句开始时取快照版本仍要取gts（当前读），也仍能保持外部一致性」。



不过在 tidb 的这篇文章中，提到了另一个时序：

1. A 开启事务并修改数据
2. B 开启事务并修改数据
3. B 提交事务
4. B 通知 A 提交
5. A 提交事务

在这个时序下，依旧存在 B 通知 A 这个因果关系，所以要保证 B 的 commit ts 一定要小于 A 的 commit ts，且如果 A 和 B 涉及的参与者完全不同时，依旧需要取一个 gts，但是对 gts 的约束变为了需要在用户发起提交到提交事务之间，此时就无法使用 start ts（或是本地缓存的 max gts）了，在发起 prepare 前或是 prepare 过程中获取可能是唯一的时机了。

说的比较啰嗦，不确定是否解答了您的问题，就作为抛砖引玉吧。欢迎随时讨论，也期待 OB 官方开发人员的回复。

**Reply by luooofan** [reply]
非常感谢！我之前没有考虑到这种时序，对几个一致性概念的理解比较粗浅。
您提到的文章也很有用，再次感谢！
